src/components/views/dialogs/SpotlightDialog.tsx

```diff
@@ -0,0 +1,544 @@
+/*
+Copyright 2021 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import React, {
+    ChangeEvent,
+    ComponentProps,
+    KeyboardEvent,
+    useCallback,
+    useContext,
+    useEffect,
+    useMemo,
+    useState,
+} from "react";
+import { Room } from "matrix-js-sdk/src/models/room";
+import { normalize } from "matrix-js-sdk/src/utils";
+import { IHierarchyRoom } from "matrix-js-sdk/src/@types/spaces";
+import { RoomHierarchy } from "matrix-js-sdk/src/room-hierarchy";
+
+import { IDialogProps } from "./IDialogProps";
+import { _t } from "../../../languageHandler";
+import BaseDialog from "./BaseDialog";
+import { BreadcrumbsStore } from "../../../stores/BreadcrumbsStore";
+import RoomAvatar from "../avatars/RoomAvatar";
+import defaultDispatcher from "../../../dispatcher/dispatcher";
+import {
+    findSiblingElement,
+    RovingAccessibleButton,
+    RovingAccessibleTooltipButton,
+    RovingTabIndexContext,
+    RovingTabIndexProvider,
+    Type,
+    useRovingTabIndex,
+} from "../../../accessibility/RovingTabIndex";
+import { Key } from "../../../Keyboard";
+import AccessibleButton from "../elements/AccessibleButton";
+import { MatrixClientPeg } from "../../../MatrixClientPeg";
+import SpaceStore from "../../../stores/spaces/SpaceStore";
+import DMRoomMap from "../../../utils/DMRoomMap";
+import { RoomType } from "matrix-js-sdk/src/@types/event";
+import { mediaFromMxc } from "../../../customisations/Media";
+import BaseAvatar from "../avatars/BaseAvatar";
+import Spinner from "../elements/Spinner";
+import { roomContextDetailsText } from "../../../Rooms";
+import DecoratedRoomAvatar from "../avatars/DecoratedRoomAvatar";
+import { Action } from "../../../dispatcher/actions";
+import Modal from "../../../Modal";
+import GenericFeatureFeedbackDialog from "./GenericFeatureFeedbackDialog";
+import AccessibleTooltipButton from "../elements/AccessibleTooltipButton";
+import RoomViewStore from "../../../stores/RoomViewStore";
+import { showStartChatInviteDialog } from "../../../RoomInvite";
+
+const MAX_RECENT_SEARCHES = 10;
+const SECTION_LIMIT = 50;
+const RECENT_SEARCHES_LS_KEY = "mx_SpotlightDialog_recent_searches";
```

Comment:
arguably should be a setting, but won't fight for it to be one here

Reply:
Good point, syncing it could be neat